features/flows/classification/src/main/java/org/opennms/netmgt/flows/classification/DefaultClassificationEngine.java

```diff
@@ -0,0 +1,117 @@
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.flows.classification;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.opennms.netmgt.flows.api.NetflowDocument;
+import org.opennms.netmgt.flows.classification.classifier.Classifier;
+import org.opennms.netmgt.flows.classification.persistence.ClassificationRuleDAO;
+import org.opennms.netmgt.flows.classification.persistence.ProtocolType;
+import org.opennms.netmgt.flows.classification.persistence.Rule;
+
+public class DefaultClassificationEngine implements ClassificationEngine {
+
+    private final ClassificationRuleDAO ruleDAO;
+
+    private final List<Classifier> classifierList = new ArrayList<>();
+
+    // This can be a DAO or similar in the future
+    public DefaultClassificationEngine(ClassificationRuleDAO ruleDAO) {
+        this.ruleDAO = Objects.requireNonNull(ruleDAO);
+        this.reload();
+    }
+
+    protected DefaultClassificationEngine(Supplier<List<Rule>> supplier) {
+        this((ClassificationRuleDAO) () -> supplier.get());
+    }
+
+    @Override
+    public String classify(NetflowDocument document) {
+        final ClassificationRequest definition = createClassificationRequest(document);
+        final String classification = classify(definition);
+        return classification;
+    }
+
+    @Override
+    public void reload() {
+        classifierList.clear();
+        ruleDAO.findAll().forEach(rule -> {
+            Classifier classifier = rule.toClassifier();
+            if (!classifierList.contains(classifier)) {
+                classifierList.add(classifier);
+            }
+        });
+        Collections.sort(classifierList, new ClassificationSorter());
+    }
+
+    @Override
+    public String classify(ClassificationRequest classificationRequest) {
+        final Optional<String> first = classifierList.stream()
+                .map(classifier -> classifier.classify(classificationRequest))
+                .filter(classifier -> classifier != null)
+                .findFirst();
+
+        // We return null instead of 'Undefined', to let the caller (e.g. rest service, or ui) decide
+        // what an unmapped definition should be named.
+        // This prevents a collision with an existing rule, which may map to 'Undefined'
+        return first.orElse(null);
+    }
+
+    protected static ClassificationRequest createClassificationRequest(NetflowDocument document) {
+        final ClassificationRequest request = new ClassificationRequest();
+        request.setProtocol(ProtocolType.createFrom(document.getIpProtocol()));
+        request.setLocation(document.getLocation());
+
+        // Decide whether to use source or dest address/port to determine application mapping
+        if (isInitiator(document)) {
```

Comment:
We should also store the result of this function within the document (as a boolean value), so that we can filter ingress/egress traffic at query time without having to reapply this logic.

Reply:
It may also be worthwhile to perform this logic only if there is a cache miss as opposed to doing it every-time there is a cache lookup.

Reply:
Yeah, I forgot about that (-: